### ISSUE TITLE
UnicodeError decoding pub key comment, when key in Putty/Pageant forwarded to host has umlaut in it

### ISSUE THREAD

**Issue by alshain** [ronf/asyncssh#118]
Thanks for the near instant feedback so far.
This time, it's probably not just user error :)

First what I did to solve my error: In Pagent (Putty's ssh-agent), I unloaded an SSH key with an umlaut in the comment.

Scenario:

Windows, running Putty/Pageant. Connected to a Ubuntu server, ssh-agent forwarding is enabled in Putty.

On Ubuntu, I run Python with asyncssh stuff. When I have a key loaded in Pageant that includes an umlaut in the comment, asyncssh is unable to decode the comment with utf-8:

https://github.com/ronf/asyncssh/blob/master/asyncssh/public_key.py#L1054

```
wait_connected
    results = await asyncio.gather(*awaitable_connections)
  File "/home/alshain/.pyenv/versions/3.6.3/lib/python3.6/asyncio/tasks.py", line 537, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/misc.py", line 159, in __await__
    return (yield from self._coro)
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/connection.py", line 4838, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/connection.py", line 4520, in create_connection
    client_keys = yield from agent.get_keys()
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/agent.py", line 251, in get_keys
    key_blob, comment))
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/agent.py", line 106, in __init__
    super().__init__(algorithm, comment)
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/public_key.py", line 1607, in __init__
    self.set_comment(comment)
  File "/home/alshain/.pyenv/versions/general/lib/python3.6/site-packages/asyncssh/public_key.py", line 1646, in set_comment
    raise KeyImportError('Invalid characters in comment') from None
asyncssh.public_key.KeyImportError: Invalid characters in comment
```

**Comment by ronf** [comment]
Non-ASCII characters are allowed in comments, but the bytes in the comment must be encoded as UTF-8 data, and not something like ISO-8859-1 (also known as "Latin 1") or Windows CP-1252.

So, for instance, a small "a"with an umlaut would be 0xe4 in Latin 1, but that would need to be encoded in the key byte string as 0xc3 0xa4 for it to be a valid UTF-8 sequence representing this same character.

The problem here is that there's no information about what encoding is used for the raw key data and the same bytes can mean different Unicode strings depending on the encoding used. So, I had to pick out. You can always read the key in yourself and do something like a key_data.decode('iso8859-1') on it before passing it on, or even convert it from its current encoding to a Unicode string and then back out to UTF-8 bytes with encode('utf-8'), if you need to write it back to a file that you want to later load into another tool like Pageant.

**Comment by alshain** [comment]
I should have clarified that I'm not doing anything except for `asyncssh.connect()`, I'm not doing any key handling by myself and I'm not trying to.

I understand that you need to guess the encoding.

Since I'm merely running `asyncssh.connect()` without anything special, I don't think that Pageant not sending UTF-8 should bring everything down.

I see multiple options:

- Don't try to parse the comment at all (unless the key is somehow queried by the user)
- Try multiple encodings, (is guessing multiple times worse than guessing once?)
- set the comment to "<decoding error>" or something like that

**Comment by ronf** [comment]
Fair point.

After thinking about this for a bit, I've reworked the handling of comment data on keys and certificates so that Unicode decoding is no longer done at the time keys are loaded into AsyncSSH. Instead, the conversion is only done when requesting the comment from a key/certificate in the form of a Unicode string, or when attempting to set a comment using a Unicode string, and those two methods have been changed to both take an optional "encoding" parameter. This parameter defaults to UTF-8 to provide backward compatibility, but now other encodings can easily be specified by a caller if the comment data is encoded in some other form.

As before, comments can be set as a byte string to avoid this issue of encoding entirely, and a new method get_comment_bytes() has been added to allow the comment data to be retrieved as a byte string without performing any Unicode decoding.

In a case like yours here where comment data is never set or retrieved, the encoding should no longer be an issue. Comment data will be kept as a byte string and never converted to a Unicode string. This should apply to all uses of keys and certificates, including those accessed through ssh-agent/Pageant.

This change is now checked into the "develop" branch if you'd like to try it.
